File: packages/sdk/src/harness/orchestrator/dependency_graph.py
"""
Dependency Graph - Workflow step dependency resolution.

This module provides the DependencyGraph class for:
- Managing step dependencies
- Detecting circular dependencies (deadlock)
- Supporting cascading skip operations
- Determining which steps are ready to execute
"""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from harness.orchestrator.types import WorkflowStep

logger = logging.getLogger(__name__)


class DependencyGraph:
    """
    Step dependency graph.

    Manages dependencies between workflow steps and supports:
    - Topological ordering for execution
    - Circular dependency (deadlock) detection
    - Cascading skip operations for conditional execution
    - Runtime state tracking (completed, skipped)

    State Management:
    - _completed: Steps that have finished (success or failure)
    - _skipped: Steps that were skipped (condition not met or dependency skipped)
    """

    def __init__(self):
        self._steps: dict[str, WorkflowStep] = {}
        self._dependencies: dict[str, set[str]] = {}
        self._completed: set[str] = set()
        self._skipped: set[str] = set()

    def add_step(self, step: WorkflowStep) -> None:
        """
        Add a step to the graph.

        Args:
            step: Workflow step to add
        """
        self._steps[step.name] = step
        if step.name not in self._dependencies:
            self._dependencies[step.name] = set()

    def add_dependency(self, step_name: str, depends_on: str) -> None:
        """
        Add a dependency relationship.

        Args:
            step_name: Step that has the dependency
            depends_on: Step that must complete first
        """
        if step_name not in self._dependencies:
            self._dependencies[step_name] = set()
        self._dependencies[step_name].add(depends_on)

    def has_pending(self) -> bool:
        """
        Check if there are pending steps.

        Returns:
            True if there are steps not yet completed or skipped
        """
        resolved = len(self._completed) + len(self._skipped)
        return resolved < len(self._steps)

    def has_only_skipped_pending(self) -> bool:
        """
        Check if all pending steps depend on skipped steps.

        This indicates we should gracefully end execution rather than
        raising a deadlock error.

        Returns:
            True if all pending steps have skipped dependencies
        """
        for name, step in self._steps.items():
            if name in self._completed or name in self._skipped:
                continue

            # Check if any dependency is not skipped
            deps = self._dependencies.get(name, set())
            if not deps or not deps.issubset(self._skipped):
                return False

        return True

    def get_ready_steps(self) -> list[WorkflowStep]:
        """
        Get steps that are ready to execute.

        A step is ready if:
        - It hasn't been completed or skipped
        - All its dependencies have been completed (not skipped)

        Note: Steps with skipped dependencies are NOT returned here,
        they should be handled by the cascade skip logic.

        Returns:
            List of steps ready for execution
        """
        ready = []
        for name, step in self._steps.items():
            if name in self._completed or name in self._skipped:
                continue

            deps = self._dependencies.get(name, set())
            skipped_deps = deps.intersection(self._skipped)

            # If any dependency was skipped, this step should be skipped too
            if skipped_deps:
                continue

            # All dependencies completed
            if deps.issubset(self._completed):
                ready.append(step)

        return ready

    def mark_completed(self, step_name: str) -> None:
        """
        Mark a step as completed.

        Args:
            step_name: Name of the completed step
        """
        self._completed.add(step_name)
        logger.debug(f"Step '{step_name}' marked as completed")

    def mark_skipped(self, step_name: str) -> None:
        """
        Mark a step as skipped and cascade to dependent steps.

        When a step is skipped, all steps that depend on it should
        also be skipped (cascade skip). This is done recursively to
        handle transitive dependencies.

        Args:
            step_name: Name of the skipped step
        """
        if step_name in self._skipped:
            return  # Already skipped

        self._skipped.add(step_name)
        logger.debug(f"Step '{step_name}' marked as skipped")

        # Find and skip all direct dependents
        dependents = self.get_dependents(step_name)
        for dependent in dependents:
            if dependent not in self._completed and dependent not in self._skipped:
                # Recursively skip this dependent and its dependents
                self.mark_skipped(dependent)

    def detect_deadlock(self) -> bool:
        """
        Detect circular dependencies (deadlock).

        Uses depth-first search to detect cycles in the dependency graph.

        Returns:
            True if a cycle is detected (deadlock)
        """
        visited = set()
        rec_stack = set()

        def has_cycle(node: str) -> bool:
            visited.add(node)
            rec_stack.add(node)

            for dep in self._dependencies.get(node, set()):
                if dep not in visited:
                    if has_cycle(dep):
                        return True
                elif dep in rec_stack:
                    logger.warning(f"Deadlock detected: cycle involving '{node}' -> '{dep}'")
                    return True

            rec_stack.remove(node)
            return False

        for step_name in self._steps:
            if step_name not in visited:
                if has_cycle(step_name):
                    return True

        return False

    def get_step(self, step_name: str) -> WorkflowStep | None:
        """Get a step by name."""
        return self._steps.get(step_name)

    def get_dependencies(self, step_name: str) -> set[str]:
        """Get dependencies for a step."""
        return self._dependencies.get(step_name, set())

    def get_dependents(self, step_name: str) -> set[str]:
        """
        Get steps that depend on the given step.

        Args:
            step_name: Step name to find dependents for

        Returns:
            Set of step names that depend on the given step
        """
        dependents = set()
        for name, deps in self._dependencies.items():
            if step_name in deps:
                dependents.add(name)
        return dependents

    def is_completed(self, step_name: str) -> bool:
        """Check if a step is completed."""
        return step_name in self._completed

    def is_skipped(self, step_name: str) -> bool:
        """Check if a step is skipped."""
        return step_name in self._skipped

    def is_resolved(self, step_name: str) -> bool:
        """Check if a step is resolved (completed or skipped)."""
        return step_name in self._completed or step_name in self._skipped

    def get_status_summary(self) -> dict:
        """
        Get a summary of the graph status.

        Returns:
            Dict with counts of steps in each state
        """
        return {
            "total": len(self._steps),
            "completed": len(self._completed),
            "skipped": len(self._skipped),
            "pending": len(self._steps) - len(self._completed) - len(self._skipped),
        }

    def reset(self) -> None:
        """Reset the graph state (keep step definitions)."""
        self._completed.clear()
        self._skipped.clear()

File: packages/sdk-java/harness-sdk-orchestrator/src/main/java/com/harness/orchestrator/DependencyGraph.java
package com.harness.orchestrator;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Step dependency graph.
 *
 * <p>Manages dependencies between workflow steps and supports:</p>
 * <ul>
 *   <li>Topological ordering for execution</li>
 *   <li>Circular dependency (deadlock) detection</li>
 *   <li>Cascading skip operations for conditional execution</li>
 *   <li>Runtime state tracking (completed, skipped)</li>
 * </ul>
 *
 * <h2>State Management</h2>
 * <ul>
 *   <li>_completed: Steps that have finished (success or failure)</li>
 *   <li>_skipped: Steps that were skipped (condition not met or dependency skipped)</li>
 * </ul>
 */
public class DependencyGraph {
    private static final Logger logger = LoggerFactory.getLogger(DependencyGraph.class);

    private final Map<String, WorkflowStep> steps = new HashMap<>();
    private final Map<String, Set<String>> dependencies = new HashMap<>();
    private final Set<String> completed = new HashSet<>();
    private final Set<String> skipped = new HashSet<>();

    /**
     * Add a step to the graph.
     */
    public void addStep(WorkflowStep step) {
        steps.put(step.getName(), step);
        if (!dependencies.containsKey(step.getName())) {
            dependencies.put(step.getName(), new HashSet<>());
        }
    }

    /**
     * Add a dependency relationship.
     *
     * @param stepName   Step that has the dependency
     * @param dependsOn  Step that must complete first
     */
    public void addDependency(String stepName, String dependsOn) {
        dependencies.computeIfAbsent(stepName, k -> new HashSet<>()).add(dependsOn);
    }

    /**
     * Check if there are pending steps.
     */
    public boolean hasPending() {
        int resolved = completed.size() + skipped.size();
        return resolved < steps.size();
    }

    /**
     * Check if all pending steps depend on skipped steps.
     *
     * <p>This indicates we should gracefully end execution rather than
     * raising a deadlock error.</p>
     */
    public boolean hasOnlySkippedPending() {
        for (Map.Entry<String, WorkflowStep> entry : steps.entrySet()) {
            String name = entry.getKey();
            if (completed.contains(name) || skipped.contains(name)) {
                continue;
            }

            Set<String> deps = dependencies.getOrDefault(name, Set.of());
            if (deps.isEmpty() || !skipped.containsAll(deps)) {
                return false;
            }
        }
        return true;
    }

    /**
     * Get steps that are ready to execute.
     *
     * <p>A step is ready if:</p>
     * <ul>
     *   <li>It hasn't been completed or skipped</li>
     *   <li>All its dependencies have been completed (not skipped)</li>
     * </ul>
     *
     * <p>Note: Steps with skipped dependencies are NOT returned here,
     * they should be handled by the cascade skip logic.</p>
     */
    public List<WorkflowStep> getReadySteps() {
        return steps.values().stream()
                .filter(step -> !completed.contains(step.getName()))
                .filter(step -> !skipped.contains(step.getName()))
                .filter(step -> {
                    Set<String> deps = dependencies.getOrDefault(step.getName(), Set.of());
                    // If any dependency was skipped, this step should be skipped too
                    if (!skipped.isEmpty() && deps.stream().anyMatch(skipped::contains)) {
                        return false;
                    }
                    // All dependencies completed
                    return completed.containsAll(deps);
                })
                .toList();
    }

    /**
     * Mark a step as completed.
     */
    public void markCompleted(String stepName) {
        completed.add(stepName);
        logger.debug("Step '{}' marked as completed", stepName);
    }

    /**
     * Mark a step as skipped and cascade to dependent steps.
     *
     * <p>When a step is skipped, all steps that depend on it should
     * also be skipped (cascade skip). This is done recursively to
     * handle transitive dependencies.</p>
     */
    public void markSkipped(String stepName) {
        if (skipped.contains(stepName)) {
            return; // Already skipped
        }

        skipped.add(stepName);
        logger.debug("Step '{}' marked as skipped", stepName);

        // Find and skip all direct dependents
        Set<String> dependents = getDependents(stepName);
        for (String dependent : dependents) {
            if (!completed.contains(dependent) && !skipped.contains(dependent)) {
                markSkipped(dependent);
            }
        }
    }

    /**
     * Detect circular dependencies (deadlock).
     *
     * <p>Uses depth-first search to detect cycles in the dependency graph.</p>
     */
    public boolean detectDeadlock() {
        Set<String> visited = new HashSet<>();
        Set<String> recStack = new HashSet<>();

        for (String stepName : steps.keySet()) {
            if (!visited.contains(stepName)) {
                if (hasCycle(stepName, visited, recStack)) {
                    return true;
                }
            }
        }
        return false;
    }

    private boolean hasCycle(String node, Set<String> visited, Set<String> recStack) {
        visited.add(node);
        recStack.add(node);

        for (String dep : dependencies.getOrDefault(node, Set.of())) {
            if (!visited.contains(dep)) {
                if (hasCycle(dep, visited, recStack)) {
                    return true;
                }
            } else if (recStack.contains(dep)) {
                logger.warn("Deadlock detected: cycle involving '{}' -> '{}'", node, dep);
                return true;
            }
        }

        recStack.remove(node);
        return false;
    }

    /**
     * Get a step by name.
     */
    public WorkflowStep getStep(String stepName) {
        return steps.get(stepName);
    }

    /**
     * Get dependencies for a step.
     */
    public Set<String> getDependencies(String stepName) {
        return new HashSet<>(dependencies.getOrDefault(stepName, Set.of()));
    }

    /**
     * Get steps that depend on the given step.
     */
    public Set<String> getDependents(String stepName) {
        Set<String> dependents = new HashSet<>();
        for (Map.Entry<String, Set<String>> entry : dependencies.entrySet()) {
            if (entry.getValue().contains(stepName)) {
                dependents.add(entry.getKey());
            }
        }
        return dependents;
    }

    /**
     * Check if a step is completed.
     */
    public boolean isCompleted(String stepName) {
        return completed.contains(stepName);
    }

    /**
     * Check if a step is skipped.
     */
    public boolean isSkipped(String stepName) {
        return skipped.contains(stepName);
    }

    /**
     * Check if a step is resolved (completed or skipped).
     */
    public boolean isResolved(String stepName) {
        return completed.contains(stepName) || skipped.contains(stepName);
    }

    /**
     * Get a summary of the graph status.
     */
    public Map<String, Integer> getStatusSummary() {
        Map<String, Integer> summary = new HashMap<>();
        summary.put("total", steps.size());
        summary.put("completed", completed.size());
        summary.put("skipped", skipped.size());
        summary.put("pending", steps.size() - completed.size() - skipped.size());
        return summary;
    }

    /**
     * Reset the graph state (keep step definitions).
     */
    public void reset() {
        completed.clear();
        skipped.clear();
    }
}
